template/interpolate/azure/keyvault/keyvault.go

```diff
@@ -0,0 +1,216 @@
+	MgmtAuthorizer := config.BearerAuthorizerCallback(sender, oauthConfig)
+	ManagementClient.Client.Authorizer = MgmtAuthorizer
+
+	return &VaultsClient, &ManagementClient, nil
+}
+
+func GetKeyVaultBaseUrlFromID(ctx context.Context, client *keyvault.VaultsClient, keyVaultId string) (string, error) {
+	if keyVaultId == "" {
+		return "", fmt.Errorf("keyVaultId is empty")
+	}
+
+	id, err := ParseAzureResourceID(keyVaultId)
+	if err != nil {
+		return "", err
+	}
+	resourceGroup := id.ResourceGroup
+
+	vaultName, ok := id.Path["vaults"]
+	if !ok {
+		return "", fmt.Errorf("resource id does not contain `vaults`: %q", keyVaultId)
+	}
+
+	resp, err := client.Get(ctx, resourceGroup, vaultName)
+	if err != nil {
+		if resp.Response.StatusCode == http.StatusNotFound {
+			return "", fmt.Errorf("Error unable to find KeyVault %q (Resource Group %q): %+v", vaultName, resourceGroup, err)
+		}
+		return "", fmt.Errorf("Error making Read request on KeyVault %q (Resource Group %q): %+v", vaultName, resourceGroup, err)
+	}
+
+	if resp.Properties == nil || resp.Properties.VaultURI == nil {
+		return "", fmt.Errorf("vault (%s) response properties or VaultURI is nil", keyVaultId)
+	}
+
+	return *resp.Properties.VaultURI, nil
+}
+
+// ParseAzureResourceID converts a long-form Azure Resource Manager ID
+// into a ResourceID. We make assumptions about the structure of URLs,
+// which is obviously not good, but the best thing available given the
+// SDK.
+func ParseAzureResourceID(id string) (*ResourceID, error) {
+	idURL, err := url.ParseRequestURI(id)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot parse Azure ID: %s", err)
+	}
+
+	path := idURL.Path
+
+	path = strings.TrimPrefix(path, "/")
+	path = strings.TrimSuffix(path, "/")
+
+	components := strings.Split(path, "/")
+
+	// We should have an even number of key-value pairs.
+	if len(components)%2 != 0 {
+		return nil, fmt.Errorf("The number of path segments is not divisible by 2 in %q", path)
+	}
+
+	var subscriptionID string
+
+	// Put the constituent key-value pairs into a map
+	componentMap := make(map[string]string, len(components)/2)
+	for current := 0; current < len(components); current += 2 {
+		key := components[current]
+		value := components[current+1]
+
+		// Check key/value for empty strings.
+		if key == "" || value == "" {
+			return nil, fmt.Errorf("Key/Value cannot be empty strings. Key: '%s', Value: '%s'", key, value)
+		}
+
+		// Catch the subscriptionID before it can be overwritten by another "subscriptions"
+		// value in the ID which is the case for the Service Bus subscription resource
+		if key == "subscriptions" && subscriptionID == "" {
+			subscriptionID = value
+		} else {
+			componentMap[key] = value
+		}
+	}
+
+	// Build up a TargetResourceID from the map
+	idObj := &ResourceID{}
+	idObj.Path = componentMap
+
+	if subscriptionID != "" {
+		idObj.SubscriptionID = subscriptionID
+	} else {
+		return nil, fmt.Errorf("No subscription ID found in: %q", path)
+	}
```

Comment:
```go
	if subscriptionID == "" {
	        return nil, fmt.Errorf("No subscription ID found in: %q", path)
	} 
        idObj.SubscriptionID = subscriptionID
```

This looks like a good opportunity to use a guard clause.

Reply:
I changed this function completely :)